### DISCUSSION TITLE
Kiwoom API 호출 횟수제한과 관련한 로직변경

### DISCUSSION THREAD

**Discussion by floatingCode21** [elbakramer/koapy#31] [Ideas]
생성해주신 좋은 프로젝트로 매매 프로그램을 만들어보고 있습니다.
Kiwoom API 호출 횟수제한과 관련한 로직을 아래와 같이 바꾸어 보았는데 문제 없을지 확인 부탁드릴게요.
초당 5회 호출제한에 대한 호출 delay를 무조건 230ms 로 고정하였고,
기준 호출횟수당 시간이 넘어가는지 확인하는 로직입니다.
기존 로직으로는 기준호출횟수가 넘어가면 무조건 긴 sleep이 들어가게 되어서,
단위시간당 매수매도 호출이 많을 경우 진행이 힘들어 보입니다.

```python
# koapy\utils\rate_limiting\RateLimiter.py
class TimeWindowRateLimiter(RateLimiter, Logging):
    def check_sleep_seconds(self):
        with self._lock:
            default_remaining = 0.23
            if len(self._call_history) < self._calls:
                return default_remaining
            else:
                clock = self._clock()
                time_limit = (self._calls / 5)
                limit_over = 0
                remaining = default_remaining
                while len(self._call_history) >= self._calls:
                    interval = clock - self._call_history[0]
                    limit_over = time_limit - interval
                    print("check_sleep_seconds : interval{}time_limit{}limit_over{}".format(interval, time_limit, limit_over))
                    if limit_over > 0:
                        remaining = self._period - limit_over
                    self._call_history.popleft()
                return remaining
```

**Comment by elbakramer** [comment]
의견 감사합니다.

덕분에 호출제한과 관련해서 살펴보다가 현재 버전에서 아래 두가지 문제가 있는걸 발견했습니다.

1. `TimeWindowRateLimiter` 의 구현이 실제 필요한 시간보다 길게 대기하도록 잘못 구현되어 있는 문제
2. 주문에 대한 제한은 초당5초 외에 추가적인 제한이 없지만 현재로는 TR 제한 프로세스에 포함되어 같이 처리되는 문제

다만 공유해주신 코드에서 `TimeWindowRateLimiter` 은 특정 제한수치와는 독립적으로 제한동작 로직만 구현하는 부분이라서요.
초당 5회 호출제한과 관련된 실제 내용은 [`KiwoomOpenApiPlusRateLimiter`](https://github.com/elbakramer/koapy/blob/master/koapy/backend/kiwoom_open_api_plus/core/KiwoomOpenApiPlusRateLimiter.py) 이쪽으로 구현하는게 맞을거 같습니다.

**Reply by floatingCode21** [reply]
빠른확인 감사합니다.
수정하신 코드 확인해보니 check_sleep_seconds 함수에서 
self._call_history.popleft() 이 들어가야 할 것으로 보입니다.

**Reply by elbakramer** [reply]
음.. 저는 `popleft()` 가 현실적으로는 굳이 필요하지 않다고 봐서 뺐어요.

왜냐면 `self._call_history = collections.deque(maxlen=self._calls)` 이기 때문에
최근 이벤트들을 최대 `self._calls` 개만 남기면서 나머지 오래된 것들은 자동으로 `popleft()` 가 되는 구조라서요.
(자세한 내용은 https://docs.python.org/3/library/collections.html#collections.deque 를 참고하세요.)

그럼에도 기존 코드에서 `popleft()` 를 추가적으로 넣었던 이유는 아래와 같은데요.
- `clock` 시점 기준으로 `self._period` 이전의 이벤트들을 굳이 들고 있을 이유가 없기 때문
- 미리 그것들을 `popleft()` 해주는 경우, 이후 체크에서 `if len(self._call_history) < self._calls:` 조건에 해당하면서 빠른 `return 0` 를 통한 성능향상 기대

근데 다시 생각해보니 `while` 문을 돌면서 일일이 하나씩 체크해서 빼는게 오히려 오버헤드가 걸리지 않을까 싶어서요.
그래서 그냥 뺐네요.

굳이 `popleft()` 동작을 집어넣는다면 아래처럼 해 볼 수는 있겠습니다.
원래 코드에서는 `remaining` 을 덮어쓰게 되면서 원래 의도하는 것보다 길게 쉬는 문제가 있었던거구요.

```python
# koapy\utils\rate_limiting\RateLimiter.py
class TimeWindowRateLimiter(RateLimiter, Logging):
    def check_sleep_seconds(self):
        with self._lock:
            if len(self._call_history) < self._calls:
                return 0
            clock = self._clock()
            remaining = self._call_history[0] + self._period - clock
            while len(self._call_history) > 0:
                remaining_per_event = self._call_history[0] + self._period - clock
                if remaining_per_event < 0:
                    self._call_history.popleft()
                else:
                    break
            return remaining
```

**Comment by floatingCode21** [comment]
아 제가 구조를 제대로 이해하지 못하고 있었네요.
친절한 설명 감사합니다.
